content/advent-2019/pkgapi.md

```diff
@@ -0,0 +1,1922 @@
+
+For example: If you're writing a library that uses an OpenGL context to do something with OpenGL, don't create the OpenGL context in the library. Instead, require the user to pass in a OpenGL context.
+
+This is also true for allocations. Where possible, don't create allocations on behalf of the user.
+
+Dave Cheney recently wrote [a most excellent article on the topic of forcing allocations](https://dave.cheney.net/2019/09/05/dont-force-allocations-on-the-callers-of-your-api). The title's a bit confusing but the main point is similar to what I am espousing here.
+
+When a library doesn't manage resources for the user, it becomes clear that the user has to manage resources by themselves. The brunt of the responsibility falls onto the user, but the library becomes more reliable.
+
+Last but not least, don't spawn goroutines on behalf of the user.
+
+
+## Easy to Use ##
+
+A good library is easy to use. There are a number of ways that a library can be easy to use.
+
+### Good Documentation and Examples ###
+
+A good library has good documentation. And to readers who think "tests are documentation", yes! Go has good support for examples, which are both documentation and tests. I enjoy using libraries that have examples when I go to their godoc.
+
+### Doesn't Panic ###
+
+Panics should only happen in a case when there are no better options. Usually returning errors are a better thing to do.
+
+### Has Minimal Dependencies ###
+
+This is fairly contentious especially in the Big Picture view of this article (see the Tension section below). But in my opinion a good library has minimal dependencies.
+
+This is especially true of libraries where source code are the primary resource being shared. If a library whose purpose is to share source code were to depend on some resource library, I would be quite suspicious.
+
+Additional dependencies also increase the difficulty to use. I often check what each library imports in order to know that my imports are not going to suddenly call home to some server somewhere. I am not fastidious over it, only because there is so much to check.
+
+### Makes the Zero Value Useful ###
+
+One of the Go proverbs, the zero value of any data type should be useful. This avoids the need for complicated constructor functions. A very good example I enjoy is Gonum's `mat.Dense` type.
+
+The `mat.Dense` data type has a method `Mul` which performs matrix multiplication. It has the following type signature:
+
+```
+func (m *Dense) Mul(a, b Matrix)
+```
+
+The result of `a × b` is placed in `m`. Thus, if `a` is a (2,3) matrix and `b` is a (3,2) matrix, then `m` will be a (2,2) matrix. The documentation is not clear, so most people will try something like this:
+
+```
+c := mat.NewDense(2, 2, make([]float64, 4))
+c.Mul(a, b)
+```
+
+
+In actuality, this would work as well:
+
+```
+var c mat.Dense
+c.Mul(a, b)
+```
+
+## Generic ##
```

Comment:
I think `Reusable` or `Reusable in different scenarios` might be a better title here. Then you don't have to explain the difference between generics.

Reply:
+1